app/ide-desktop/lib/dashboard/src/hooks/eventCallback.ts

```diff
@@ -0,0 +1,25 @@
+/**
+ * @file useEventCallback shim from the latest React RFC.
+ */
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import * as React from 'react'
+
+import * as syncRef from './syncRef'
+
+/**
+ * useEventCallback shim from the latest React RFC.
+ * @see https://github.com/reactjs/rfcs/pull/220
+ * @see https://github.com/reactjs/rfcs/blob/useevent/text/0000-useevent.md#internal-implementation
+ */
+export function useEventCallback<
+  Func extends (...args: Args) => Result,
```

Comment:
why are we using three separate generic parameters here btw? (to reconstruct a type we already have, as well!)

(and it loses generics and overloads too D:)

Reply:
Because often I had an experience when I have a function type, and I want to express: wrap that function with the same signature.

Reply:
Overloads also doesn't make sense here, this function is too simple to have overloads. Does `useCallback` have them? :)

Reply:
i mean if the callback has overloads they will disappear ._.

Reply:
[playground repro for both cases.](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtX2QGcQBRANxFQwGEoIIAjKMAawB4AoeeAMTTHggAHhkrBC8ABQA6WbADmhAFzwAgjEUBKeAF4AfPABKIQsggYANFzUaJw0anHwACrCgBbEKJiE2UVACeBjoubp7evnzoelbcxqbmgiJiEsYYyDCoACoBAA4gfoHBRl4Z2XkFUWAxHHqSYHSMzCwqVZoqMnK2KupaugbxZhgA3BwcYHiEGPAA7gAWUNMhRKQUVLT0TKySbFkWdVAqWdr68FCaowD0l9zcAHoA-GOgkLAIKOjYeEiSh-CoyHcDBAMHa-0BwJgHBe0DgSH4X3wiF+KimMCwqHkYLRGPkYwmqCmswWGAATLoCMRyJQaI0tixJIgLhxrrd4I8xvNFqTJABGZlcsmSADkUGFzKAA)

```ts
declare function useEventCallback<
  Func extends (...args: Args) => Result,
  Args extends Parameters<any> = Parameters<Func>,
  Result extends ReturnType<any> = ReturnType<Func>,
>(callback: Func): (...args: Args) => Result;

const what = useEventCallback(<T,>(a: T) => a);
//    ^?

declare function f(a: number): number
declare function f(a: string): string

const what2 = useEventCallback(f);
//    ^?

what2(1);
what2('a');
```

Reply:
IMO, it's a limitation in typescript, not in the declaration. I'm not sure if there's a way to workaroud this except manually casting types.

If you've found a better solution, I'd be happy if you share it with me :)

Reply:
[playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABCAzgUwKIDc1igYQEMAbYgI0IgGsAeAKEUQDFwJE0APKXAExUQAUAOhGEATgHMUALkSEwATwDaAXQCUiALwA+ZGCpg4AdzB1tAiCXKUqslpDV3WiAN4NEAeg+IAKgAsYfigFAAc0ORR0MVgEREDEAGUAQSYMABoIuKg4-kJEKDF5FBDxXGyjQpCwsTkxOHAefL9wmDAwNBrQSBiwIXcxNCgQMSQBYVFJFA0dREtSCmohOeIBMBBSDPEpNQ1CfnsIAG46AF86OggEFHK-QmzNZHRsMqJ5mwEaHzTzQlkfad0hDUxy8jEYAD0APwXK5wYhoITEOASARGW5QAQARh2IO8YPxEOhlzAKDhCKRKLRdwEAHJCDScXRQQT8VCYSSyYjkaj0QIXCdGcyWYTzl1oPAkMABL9EGsALZkDqOWUgBUdOhinqIKUy65iVoSZV6g0a1hanWyeWKmoAH0QxrAhstqutiDtDokrn6g2GSGCYTgwDkWk0DxpVo6NMQkODAGpEJjELI8vHCKdzsTrogqVAAExaR6YHB4V7WagCYDApl4kXs0nwrmU9G5rGCmvCtmZzkUnl3Ft0hlVoUd6FAA)

```ts
function useEventCallback<
  Func extends (...args: any[]) => unknown
>(callback: Func): Func {
  // React hooks stripped out because the playground does't have React.
  // It will work the same with hooks.
  return ((...args) => callback.call(null, args)) as Func;
}
```

technically it requires a type assertion. but it's one assertion (= in one place) that is known to be safe

Reply:
Yea, this should work. Will update tomorrow, thanks :)